### PR TITLE
Bring amp in line with canonical definition

### PR DESCRIPTION
**Pull request by WT-MM**
Use the canonical AMP reward
Add gradient penalty for discriminator objective

### CODE REVIEW

**Review by codekansas — COMMENTED** [review]
wait nvm. looking at this closer

File: ksim/task/amp.py [hunk 1]
```diff
@@ -357,6 +361,30 @@ def get_disc_losses(
...
+        disc_model: PyTree,
+        real_motions: PyTree,  # (B, T, ...)
+        rng: PRNGKeyArray,
+    ) -> Array:
+        """R1 gradient penalty  (w.r.t. the *input* features)."""
+
+        def disc_on_motion(motion: PyTree, key: PRNGKeyArray) -> Array:
+            return self.call_discriminator(disc_model, motion, key).sum()
+
+        grad_fn = jax.grad(disc_on_motion, argnums=0)
```
Comment (codekansas) [inline ksim/task/amp.py:375]: okay, this part really confused me. `real_motions` is not parametrized, so applying gradients to it doesn't make any sense - you're basically trying to "update" the real data, which is not what you want to do.

i have never seen the trick in the paper, but it looks like the idea is to minimize the discriminator gradients on the real samples. it should be something like this

```suggestion
        def disc_on_motion(disc_model: PyTree, motion: PyTree, key: PRNGKeyArray) -> Array:
            return self.call_discriminator(disc_model, motion, key).sum()

        grad_fn = jax.grad(disc_on_motion, argnums=0)
```

then you need to change how they're applied below as well

**Review by codekansas — CHANGES_REQUESTED** [review]

File: ksim/task/amp.py [hunk 2]
```diff
@@ -75,6 +75,10 @@ class AMPConfig(PPOConfig):
         value=32,
         help="The batch size for reference motion batching in AMP training.",
     )
+    amp_grad_penalty_coef: float = xax.field(
+        value=0.5,
```
Comment (codekansas) [inline ksim/task/amp.py]: in the paper this value is 10, and it gets divided by 2 in the equation

**Review by codekansas — APPROVED** [review]
ok, after staring at this more, seems legit
